File: perl/test/test_roundtrip.pl
#!/usr/bin/env perl
#
# test_roundtrip.pl - Bidirectional testing with round-trip verification
#
# Tests:
# 1. Pack direction (unpacked -> packed) with timing
# 2. Unpack direction (packed -> unpacked) with timing
# 3. Unpacked round-trip: unpack(pack(x)) = x
# 4. Packed round-trip: pack(unpack(y)) = y
#
# Usage: perl test_roundtrip.pl <csv_file>
#

use strict;
use warnings;
use Time::HiRes qw(time);
use FindBin qw($Bin);
use lib "$Bin/../src";
use MPC::Designation qw(convert_simple);

my $MAX_ERRORS = 20;

sub main {
    die "Usage: $0 <csv_file>\n" unless @ARGV >= 1;
    my $csv_file = $ARGV[0];

    # Load test data
    open my $fh, '<', $csv_file or die "Cannot open $csv_file: $!";
    my $header = <$fh>;  # Skip header

    my @unpacked_list;
    my @packed_list;

    while (my $line = <$fh>) {
        chomp $line;
        next unless $line;
        my @parts = split /,/, $line, 3;
        next unless @parts >= 2;
        push @unpacked_list, $parts[0];
        push @packed_list, $parts[1];
    }
    close $fh;

    my $total = scalar @unpacked_list;
    print "Loaded $total test cases\n\n";

    my @errors;

    # ========== Phase 1: Pack (unpacked -> packed) ==========
    print "=== Phase 1: Pack (unpacked -> packed) ===\n";
    my $pack_passed = 0;
    my $pack_failed = 0;
    my $start = time();

    for my $i (0 .. $total - 1) {
        my $unpacked = $unpacked_list[$i];
        my $expected = $packed_list[$i];

        my $got = eval { convert_simple($unpacked) };
        if ($@) {
            $pack_failed++;
            push @errors, ['pack', $unpacked, "ERROR: $@", $expected] if @errors < $MAX_ERRORS;
        } elsif ($got ne $expected) {
            $pack_failed++;
            push @errors, ['pack', $unpacked, $got, $expected] if @errors < $MAX_ERRORS;
        } else {
            $pack_passed++;
        }
    }

    my $pack_time = int((time() - $start) * 1000);
    my $pack_rate = $total / ($pack_time / 1000);
    print "Passed: $pack_passed\n";
    print "Failed: $pack_failed\n";
    printf "Time:   %dms (%.1f entries/sec)\n\n", $pack_time, $pack_rate;

    # ========== Phase 2: Unpack (packed -> unpacked) ==========
    print "=== Phase 2: Unpack (packed -> unpacked) ===\n";
    my $unpack_passed = 0;
    my $unpack_failed = 0;
    $start = time();

    for my $i (0 .. $total - 1) {
        my $packed = $packed_list[$i];
        my $expected = $unpacked_list[$i];

        my $got = eval { convert_simple($packed) };
        if ($@) {
            $unpack_failed++;
            push @errors, ['unpack', $packed, "ERROR: $@", $expected] if @errors < $MAX_ERRORS;
        } elsif ($got ne $expected) {
            $unpack_failed++;
            push @errors, ['unpack', $packed, $got, $expected] if @errors < $MAX_ERRORS;
        } else {
            $unpack_passed++;
        }
    }

    my $unpack_time = int((time() - $start) * 1000);
    my $unpack_rate = $total / ($unpack_time / 1000);
    print "Passed: $unpack_passed\n";
    print "Failed: $unpack_failed\n";
    printf "Time:   %dms (%.1f entries/sec)\n\n", $unpack_time, $unpack_rate;

    # ========== Phase 3: Unpacked round-trip: unpack(pack(x)) = x ==========
    print "=== Phase 3: Unpacked round-trip: unpack(pack(x)) = x ===\n";
    my $rt_unpacked_passed = 0;
    my $rt_unpacked_failed = 0;
    $start = time();

    for my $i (0 .. $total - 1) {
        my $original = $unpacked_list[$i];

        my ($packed, $unpacked);
        eval {
            $packed = convert_simple($original);
            $unpacked = convert_simple($packed);
        };
        if ($@) {
            $rt_unpacked_failed++;
            next;
        }

        if ($unpacked ne $original) {
            $rt_unpacked_failed++;
            push @errors, ['rt-unp', $original, "$packed -> $unpacked", $original] if @errors < $MAX_ERRORS;
        } else {
            $rt_unpacked_passed++;
        }
    }

    my $rt_unpacked_time = int((time() - $start) * 1000);
    my $rt_unpacked_rate = $total / ($rt_unpacked_time / 1000);
    print "Passed: $rt_unpacked_passed\n";
    print "Failed: $rt_unpacked_failed\n";
    printf "Time:   %dms (%.1f entries/sec)\n\n", $rt_unpacked_time, $rt_unpacked_rate;

    # ========== Phase 4: Packed round-trip: pack(unpack(y)) = y ==========
    print "=== Phase 4: Packed round-trip: pack(unpack(y)) = y ===\n";
    @errors = ();  # Reset to show phase 4 errors
    my $rt_packed_passed = 0;
    my $rt_packed_failed = 0;
    $start = time();

    for my $i (0 .. $total - 1) {
        my $original = $packed_list[$i];

        my ($unpacked, $repacked);
        eval {
            $unpacked = convert_simple($original);
            $repacked = convert_simple($unpacked);
        };
        if ($@) {
            $rt_packed_failed++;
            push @errors, ['rt-pak', $original, "ERROR: $@", $original] if @errors < $MAX_ERRORS;
            next;
        }

        if ($repacked ne $original) {
            $rt_packed_failed++;
            push @errors, ['rt-pak', $original, "$unpacked -> $repacked", $original] if @errors < $MAX_ERRORS;
        } else {
            $rt_packed_passed++;
        }
    }

    my $rt_packed_time = int((time() - $start) * 1000);
    my $rt_packed_rate = $total / ($rt_packed_time / 1000);
    print "Passed: $rt_packed_passed\n";
    print "Failed: $rt_packed_failed\n";
    printf "Time:   %dms (%.1f entries/sec)\n\n", $rt_packed_time, $rt_packed_rate;

    # ========== Summary ==========
    print "=== Summary ===\n";
    printf "%-30s %10s %18s %12s\n", "Phase", "Time (ms)", "Rate (entries/sec)", "Status";
    printf "%-30s %10s %18s %12s\n", "-" x 30, "-" x 10, "-" x 18, "-" x 12;

    sub status {
        my $failed = shift;
        return "PASS" if $failed == 0;
        return "FAIL ($failed)";
    }

    printf "%-30s %10d %18.1f %12s\n", "Pack", $pack_time, $pack_rate, status($pack_failed);
    printf "%-30s %10d %18.1f %12s\n", "Unpack", $unpack_time, $unpack_rate, status($unpack_failed);
    printf "%-30s %10d %18.1f %12s\n", "Unpacked RT: unpack(pack(x))=x", $rt_unpacked_time, $rt_unpacked_rate, status($rt_unpacked_failed);
    printf "%-30s %10d %18.1f %12s\n", "Packed RT: pack(unpack(y))=y", $rt_packed_time, $rt_packed_rate, status($rt_packed_failed);
    print "\n";

    # Show errors
    if (@errors) {
        print "=== First " . scalar(@errors) . " errors ===\n";
        printf "%-8s %-25s %-20s %-20s\n", "Phase", "Input", "Got", "Expected";
        printf "%-8s %-25s %-20s %-20s\n", "-" x 8, "-" x 25, "-" x 20, "-" x 20;
        for my $e (@errors) {
            printf "%-8s %-25s %-20s %-20s\n", @$e;
        }
    }

    # Exit with error only if pack or packed RT failed
    my $total_failed = $pack_failed + $rt_packed_failed;
    exit($total_failed > 0 ? 1 : 0);
}

main();

File: python/test/test_roundtrip.py
#!/usr/bin/env python3
"""
Test MPC designation conversion with bidirectional timing and round-trip verification.

Tests:
1. Pack direction (unpacked -> packed) with timing
2. Unpack direction (packed -> unpacked) with timing
3. Unpacked round-trip: unpack(pack(x)) = x
4. Packed round-trip: pack(unpack(y)) = y

Usage: python test_roundtrip.py <csv_file>
"""

import sys
import time
import os

# Add parent src directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))
from mpc_designation import convert_simple

MAX_ERRORS = 20


def main():
    if len(sys.argv) < 2:
        print("Usage: python test_roundtrip.py <csv_file>", file=sys.stderr)
        sys.exit(1)

    csv_file = sys.argv[1]

    # Load test data
    unpacked_list = []
    packed_list = []

    with open(csv_file, 'r') as f:
        next(f)  # Skip header
        for line in f:
            line = line.strip()
            if not line:
                continue
            parts = line.split(',', 2)
            if len(parts) >= 2:
                unpacked_list.append(parts[0])
                packed_list.append(parts[1])

    total = len(unpacked_list)
    print(f"Loaded {total} test cases\n")

    errors = []

    # ========== Phase 1: Pack (unpacked -> packed) ==========
    print("=== Phase 1: Pack (unpacked -> packed) ===")
    pack_passed = 0
    pack_failed = 0
    start = time.time()

    for i in range(total):
        unpacked = unpacked_list[i]
        expected = packed_list[i]
        try:
            got = convert_simple(unpacked)
            if got == expected:
                pack_passed += 1
            else:
                pack_failed += 1
                if len(errors) < MAX_ERRORS:
                    errors.append(('pack', unpacked, got, expected))
        except Exception as e:
            pack_failed += 1
            if len(errors) < MAX_ERRORS:
                errors.append(('pack', unpacked, f'ERROR: {e}', expected))

    pack_time = (time.time() - start) * 1000
    pack_rate = total / (pack_time / 1000)
    print(f"Passed: {pack_passed}")
    print(f"Failed: {pack_failed}")
    print(f"Time:   {pack_time:.0f}ms ({pack_rate:.1f} entries/sec)\n")

    # ========== Phase 2: Unpack (packed -> unpacked) ==========
    print("=== Phase 2: Unpack (packed -> unpacked) ===")
    unpack_passed = 0
    unpack_failed = 0
    start = time.time()

    for i in range(total):
        packed = packed_list[i]
        expected = unpacked_list[i]
        try:
            got = convert_simple(packed)
            if got == expected:
                unpack_passed += 1
            else:
                unpack_failed += 1
                if len(errors) < MAX_ERRORS:
                    errors.append(('unpack', packed, got, expected))
        except Exception as e:
            unpack_failed += 1
            if len(errors) < MAX_ERRORS:
                errors.append(('unpack', packed, f'ERROR: {e}', expected))

    unpack_time = (time.time() - start) * 1000
    unpack_rate = total / (unpack_time / 1000)
    print(f"Passed: {unpack_passed}")
    print(f"Failed: {unpack_failed}")
    print(f"Time:   {unpack_time:.0f}ms ({unpack_rate:.1f} entries/sec)\n")

    # ========== Phase 3: Unpacked round-trip: unpack(pack(x)) = x ==========
    print("=== Phase 3: Unpacked round-trip: unpack(pack(x)) = x ===")
    rt_unpacked_passed = 0
    rt_unpacked_failed = 0
    start = time.time()

    for i in range(total):
        original = unpacked_list[i]
        try:
            packed = convert_simple(original)
            unpacked = convert_simple(packed)
            if unpacked == original:
                rt_unpacked_passed += 1
            else:
                rt_unpacked_failed += 1
                if len(errors) < MAX_ERRORS:
                    errors.append(('rt-unp', original, f'{packed} -> {unpacked}', original))
        except Exception as e:
            rt_unpacked_failed += 1

    rt_unpacked_time = (time.time() - start) * 1000
    rt_unpacked_rate = total / (rt_unpacked_time / 1000)
    print(f"Passed: {rt_unpacked_passed}")
    print(f"Failed: {rt_unpacked_failed}")
    print(f"Time:   {rt_unpacked_time:.0f}ms ({rt_unpacked_rate:.1f} entries/sec)\n")

    # ========== Phase 4: Packed round-trip: pack(unpack(y)) = y ==========
    print("=== Phase 4: Packed round-trip: pack(unpack(y)) = y ===")
    errors = []  # Reset to show phase 4 errors
    rt_packed_passed = 0
    rt_packed_failed = 0
    start = time.time()

    for i in range(total):
        original = packed_list[i]
        try:
            unpacked = convert_simple(original)
            repacked = convert_simple(unpacked)
            if repacked == original:
                rt_packed_passed += 1
            else:
                rt_packed_failed += 1
                if len(errors) < MAX_ERRORS:
                    errors.append(('rt-pak', original, f'{unpacked} -> {repacked}', original))
        except Exception as e:
            rt_packed_failed += 1
            if len(errors) < MAX_ERRORS:
                errors.append(('rt-pak', original, f'ERROR: {e}', original))

    rt_packed_time = (time.time() - start) * 1000
    rt_packed_rate = total / (rt_packed_time / 1000)
    print(f"Passed: {rt_packed_passed}")
    print(f"Failed: {rt_packed_failed}")
    print(f"Time:   {rt_packed_time:.0f}ms ({rt_packed_rate:.1f} entries/sec)\n")

    # ========== Summary ==========
    print("=== Summary ===")
    print(f"{'Phase':<30} {'Time (ms)':>10} {'Rate (entries/sec)':>18} {'Status':>12}")
    print(f"{'-'*30} {'-'*10} {'-'*18} {'-'*12}")

    def status(failed):
        return "PASS" if failed == 0 else f"FAIL ({failed})"

    print(f"{'Pack':<30} {pack_time:>10.0f} {pack_rate:>18.1f} {status(pack_failed):>12}")
    print(f"{'Unpack':<30} {unpack_time:>10.0f} {unpack_rate:>18.1f} {status(unpack_failed):>12}")
    print(f"{'Unpacked RT: unpack(pack(x))=x':<30} {rt_unpacked_time:>10.0f} {rt_unpacked_rate:>18.1f} {status(rt_unpacked_failed):>12}")
    print(f"{'Packed RT: pack(unpack(y))=y':<30} {rt_packed_time:>10.0f} {rt_packed_rate:>18.1f} {status(rt_packed_failed):>12}")
    print()

    # Show errors
    if errors:
        print(f"=== First {len(errors)} errors ===")
        print(f"{'Phase':<8} {'Input':<25} {'Got':<20} {'Expected':<20}")
        print(f"{'-'*8} {'-'*25} {'-'*20} {'-'*20}")
        for phase, inp, got, expected in errors:
            print(f"{phase:<8} {inp:<25} {got:<20} {expected:<20}")

    # Exit with error only if pack or packed RT failed (those are real bugs)
    total_failed = pack_failed + rt_packed_failed
    sys.exit(1 if total_failed > 0 else 0)


if __name__ == '__main__':
    main()
